### PR TITLE
Modularize definition and use of typed errors

### PR DESCRIPTION
**Pull request by tschneidereit**
This introduces support for adding typed error definitions in a modular way, such that extensions can define their own typed errors.

This support is then also used in a few places, though more work can be done there.

### CODE REVIEW

File: builtins/web/fetch/fetch-errors.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+DEF_ERR(InvalidCtorInitArg, JSEXN_TYPEERR, "{0} constructor: |init| parameter can't be converted to a dictionary", 1)
+DEF_ERR(NonBodyRequestWithBody, JSEXN_TYPEERR, "Request constructor: HEAD or GET Request cannot have a body", 0)
+DEF_ERR(NonBodyResponseWithBody, JSEXN_TYPEERR, "Response with status {0} cannot have body", 1)
+DEF_ERR(BodyStreamUnusable, JSEXN_TYPEERR, "Can't use a ReadableStream that's locked or has ever been read from or canceled", 0)
+DEF_ERR(InvalidStatus, JSEXN_TYPEERR, "Response constructor: invalid status {0}", 1)
+DEF_ERR(InvalidStreamChunk, JSEXN_TYPEERR, "ReadableStream used as a Request or Response body must produce Uint8Array values", 0)
+DEF_ERR(EmptyHeaderName, JSEXN_TYPEERR, "{0}: Header name can't be empty", 1)
+DEF_ERR(InvalidHeaderName, JSEXN_TYPEERR, "{0}: Invalid header name \"{1}\"", 2)
+DEF_ERR(InvalidHeaderValue, JSEXN_TYPEERR, "{0}: Invalid header value \"{1}\"", 2)
+DEF_ERR(HeadersCloningFailed, JSEXN_INTERNALERR, "Failed to clone headers", 0)
```
Comment (guybedford) [inline builtins/web/fetch/fetch-errors.h:19]: `InternalError` is not really used in JS - can we just use a normal `JSEXN_ERR` instead?

File: include/error-numbers.msg [hunk 2]
```diff
@@ -47,20 +47,16 @@ MSG_DEF(JSMSG_INCOMPATIBLE_INSTANCE,                           2, JSEXN_TYPEERR,
...
-MSG_DEF(JSMSG_RESPONSE_NULL_BODY_STATUS_WITH_BODY,             0, JSEXN_TYPEERR, "Response with null body status cannot have body")
 MSG_DEF(JSMSG_RESPONSE_JSON_INVALID_VALUE,                     0, JSEXN_TYPEERR, "Redirect.json: The data is not JSON serializable")
 MSG_DEF(JSMSG_TEXT_DECODER_INVALID_ENCODING,                   1, JSEXN_RANGEERR, "TextDecoder constructor: The given encoding '{0}' is not supported.")
 MSG_DEF(JSMSG_TEXT_DECODER_DECODING_FAILED,                    0, JSEXN_TYPEERR, "TextDecoder.decode: Decoding failed.")
 MSG_DEF(JSMSG_TEXT_DECODER_OPTIONS_NOT_DICTIONARY,             0, JSEXN_TYPEERR, "TextDecoder constructor: options argument can't be converted to a dictionary.")
 MSG_DEF(JSMSG_TEXT_DECODER_DECODE_OPTIONS_NOT_DICTIONARY,      0, JSEXN_TYPEERR, "TextDecoder.decode: options argument can't be converted to a dictionary.")
 MSG_DEF(JSMSG_TEXT_ENCODER_ENCODEINTO_INVALID_ARRAY,           0, JSEXN_TYPEERR, "TextEncoder.encodeInto: Argument 2 does not implement interface Uint8Array.")
-//clang-format on
\ No newline at end of file
+//clang-format on
```
Comment (guybedford) [inline include/error-numbers.msg:62]: Will this new error system ultimately replace `include/error-numbers.msg` as a design then? Or are there use cases for both still?

File: include/errors.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+                 const char* arg3 = nullptr,
+                 const char* arg4 = nullptr);
+
+namespace Errors {
+DEF_ERR(InvalidSequence, JSEXN_TYPEERR, "Failed to construct {0} object. If defined, the first "
+                     "argument must be either a [ ['name', 'value'], ... ] sequence, "
+                     "or a { 'name' : 'value', ... } record{1}.", 2)
+DEF_ERR(ForEachCallback, JSEXN_TYPEERR, "Failed to execute 'forEach' on '{0}': "
+                                        "parameter 1 is not of type 'Function'", 1)
+DEF_ERR(RequestHandlerOnly, JSEXN_EVALERR, "{0} can only be used during request handling, "                             \
```
Comment (guybedford) [inline include/errors.h:16]: `EvalError` is not really used in JS at all either, again can we just use a normal `Error` or `TypeError`?

File: include/errors.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+                                           "not during initialization", 1)
+DEF_ERR(InitializationOnly, JSEXN_EVALERR, "{0} can only be used during request handling, "
```
Comment (guybedford) [inline include/errors.h:18]: And here.

**Review by guybedford — APPROVED** [review]

File: include/error-numbers.msg (hunk 2 above)
Comment (tschneidereit) [reply include/error-numbers.msg:62]: The idea is for it to replace `error-numbers.msg`, yes. It's much easier to use in a modular way, so I don't really see a reason to keep the two systems around. But please let me know if you see one!

Comment (guybedford) [reply include/error-numbers.msg:62]: Sounds good to me, just wanted to verify.

**Pull request merged**